File: app/src/main/java/com/glookogluroo/bridge/glooko/GlookoParser.kt
package com.glookogluroo.bridge.glooko

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.double
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter

object GlookoParser {
    private val json = Json { ignoreUnknownKeys = true }

    private val glookoLocalDateTimeFormatters = listOf(
        DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS"),
        DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss"),
        DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"),
    )

    fun parseTimestamp(value: JsonElement?, zone: ZoneId = ZoneId.systemDefault()): Instant? {
        if (value == null) return null
        return when {
            value.jsonPrimitive.isString -> parseTimestampString(value.jsonPrimitive.content, zone)
            value.jsonPrimitive.content.toDoubleOrNull() != null -> {
                val epoch = value.jsonPrimitive.double
                if (epoch > 1e12) Instant.ofEpochMilli(epoch.toLong())
                else Instant.ofEpochSecond(epoch.toLong())
            }
            else -> null
        }
    }

    fun parseTimestamp(value: String?, zone: ZoneId = ZoneId.systemDefault()): Instant? {
        if (value.isNullOrBlank()) return null
        return parseTimestampString(value, zone)
    }

    private fun parseTimestampString(value: String, zone: ZoneId): Instant? {
        // Glooko pump graph timestamps are local wall-clock times with a literal "Z" suffix,
        // not true UTC. Match glooko-reader's naive-datetime parsing, then apply the account zone.
        if (value.endsWith("Z", ignoreCase = true)) {
            parseLocalDateTime(value.dropLast(1).trim())?.let { local ->
                return local.atZone(zone).toInstant()
            }
        }

        parseLocalDateTime(value)?.let { local ->
            return local.atZone(zone).toInstant()
        }

        runCatching {
            return Instant.parse(value)
        }

        return null
    }

    private fun parseLocalDateTime(value: String): LocalDateTime? {
        for (formatter in glookoLocalDateTimeFormatters) {
            runCatching {
                return LocalDateTime.parse(value, formatter)
            }
        }
        runCatching {
            return LocalDateTime.parse(value, DateTimeFormatter.ISO_LOCAL_DATE_TIME)
        }
        return null
    }

    fun parseBolusEntries(graphData: JsonObject): List<BolusEntry> {
        val series = graphData["series"]?.jsonObject ?: graphData
        val deliveredBolus = series["deliveredBolus"]?.jsonArray ?: return emptyList()

        return deliveredBolus.mapNotNull { element ->
            val bolus = element.jsonObject
            val timestamp = parseTimestamp(bolus["timestamp"]) ?: parseTimestamp(bolus["x"])
                ?: return@mapNotNull null

            val units = bolus["insulinDelivered"]?.jsonPrimitive?.content?.toDoubleOrNull()
                ?: bolus["y"]?.jsonPrimitive?.content?.toDoubleOrNull()
                ?: bolus["value"]?.jsonPrimitive?.content?.toDoubleOrNull()
                ?: return@mapNotNull null

            if (units <= 0) return@mapNotNull null

            BolusEntry(
                timestamp = timestamp,
                units = units,
                carbsInput = bolus["carbsInput"]?.jsonPrimitive?.content?.toDoubleOrNull(),
                insulinOnBoard = bolus["insulinOnBoard"]?.jsonPrimitive?.content?.toDoubleOrNull(),
                bloodGlucoseInput = bolus["bloodGlucoseInput"]?.jsonPrimitive?.content?.toIntOrNull(),
                bloodGlucoseSource = bolus["bloodGlucoseInputSource"]?.jsonPrimitive?.content,
                correctionUnits = bolus["insulinRecommendationForCorrection"]?.jsonPrimitive?.content?.toDoubleOrNull(),
                carbUnits = bolus["insulinRecommendationForCarbs"]?.jsonPrimitive?.content?.toDoubleOrNull(),
                isManual = bolus["isManual"]?.jsonPrimitive?.content?.toBooleanStrictOrNull() ?: false,
                deviceName = bolus["deviceName"]?.jsonPrimitive?.content,
            )
        }
    }

    fun parsePumpMode(statistics: JsonObject): PumpStatistics? {
        val autoPct = statistics["op5PumpModeAutomaticPercentage"]?.jsonPrimitive?.content?.toDoubleOrNull()
        val manualPct = statistics["op5PumpModeManualPercentage"]?.jsonPrimitive?.content?.toDoubleOrNull()
        val limitedPct = statistics["op5PumpModeLimitedPercentage"]?.jsonPrimitive?.content?.toDoubleOrNull()
        val hypoPct = statistics["op5PumpModeHypoProtectPercentage"]?.jsonPrimitive?.content?.toDoubleOrNull()

        if (autoPct != null || manualPct != null) {
            val auto = autoPct ?: 0.0
            val manual = manualPct ?: 0.0
            val limited = limitedPct ?: 0.0
            val mode = when {
                auto >= manual && auto >= limited -> PumpMode.AUTO
                manual >= limited -> PumpMode.MANUAL
                else -> PumpMode.LIMITED
            }
            return PumpStatistics(
                mode = mode,
                autoPercentage = auto,
                manualPercentage = manual,
                limitedPercentage = limited,
                hypoProtectPercentage = hypoPct ?: 0.0,
                totalInsulinPerDay = statistics["totalInsulinPerDay"]?.jsonPrimitive?.content?.toDoubleOrNull(),
                basalPercentage = statistics["basalPercentage"]?.jsonPrimitive?.content?.toDoubleOrNull(),
                bolusPercentage = statistics["bolusPercentage"]?.jsonPrimitive?.content?.toDoubleOrNull(),
                averageBg = statistics["averageBg"]?.jsonPrimitive?.content?.toDoubleOrNull(),
                inRangePercentage = statistics["inRangePercentage"]?.jsonPrimitive?.content?.toDoubleOrNull(),
                carbsPerDay = statistics["carbsPerDay"]?.jsonPrimitive?.content?.toDoubleOrNull(),
            )
        }

        val pumpModes = statistics["pumpModes"]?.jsonObject
            ?: statistics["pump_modes"]?.jsonObject
            ?: return null

        val auto = pumpModes["auto"]?.jsonPrimitive?.content?.toDoubleOrNull()
            ?: pumpModes["automated"]?.jsonPrimitive?.content?.toDoubleOrNull()
            ?: 0.0
        val manual = pumpModes["manual"]?.jsonPrimitive?.content?.toDoubleOrNull() ?: 0.0
        val limited = pumpModes["limited"]?.jsonPrimitive?.content?.toDoubleOrNull() ?: 0.0
        val mode = when {
            auto >= manual && auto >= limited -> PumpMode.AUTO
            manual >= limited -> PumpMode.MANUAL
            else -> PumpMode.LIMITED
        }

        return PumpStatistics(
            mode = mode,
            autoPercentage = auto,
            manualPercentage = manual,
            limitedPercentage = limited,
        )
    }

    fun parseDevices(deviceData: JsonObject): List<DeviceInfo> {
        val devices = deviceData["devices"]?.jsonArray ?: return emptyList()
        return devices.mapNotNull { element ->
            val dev = element.jsonObject
            val properties = dev["properties"]?.jsonObject?.entries?.associate { (key, value) ->
                key to value.jsonPrimitive.content
            } ?: emptyMap()

            DeviceInfo(
                name = dev["displayName"]?.jsonPrimitive?.content
                    ?: dev["shortDisplayName"]?.jsonPrimitive?.content
                    ?: "Unknown",
                brand = dev["brand"]?.jsonPrimitive?.content,
                model = dev["model"]?.jsonPrimitive?.content,
                serialNumber = dev["serialNumber"]?.jsonPrimitive?.content,
                deviceType = dev["type"]?.jsonPrimitive?.content,
                lastSync = parseTimestamp(dev["lastSyncTimestamp"]),
                properties = properties,
            )
        }
    }

    fun parseJsonObject(raw: String): JsonObject = json.parseToJsonElement(raw).jsonObject
}

File: backend/g2g/glooko_parser.py
"""Glooko JSON parsers — ported from GlookoParser.kt."""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any
from zoneinfo import ZoneInfo

from g2g.models import BolusEntry, DeviceInfo, PumpMode, PumpStatistics


def parse_timestamp(value: Any, zone: ZoneInfo | None = None) -> datetime | None:
    """Parse Glooko timestamps.

    Strings ending in Z are treated as local wall-clock (not UTC).
    """
    if value is None:
        return None
    zone = zone or ZoneInfo("America/Los_Angeles")

    if isinstance(value, (int, float)):
        epoch = float(value)
        if epoch > 1e12:
            return datetime.fromtimestamp(epoch / 1000.0, tz=timezone.utc)
        return datetime.fromtimestamp(epoch, tz=timezone.utc)

    text = str(value).strip()
    if not text:
        return None

    # Numeric string
    try:
        return parse_timestamp(float(text), zone)
    except ValueError:
        pass

    local_text = text
    treat_as_local = False
    if text.endswith("Z") or text.endswith("z"):
        local_text = text[:-1]
        treat_as_local = True

    formats = (
        "%Y-%m-%dT%H:%M:%S.%f",
        "%Y-%m-%dT%H:%M:%S",
        "%Y-%m-%d %H:%M:%S",
    )
    for fmt in formats:
        try:
            naive = datetime.strptime(local_text, fmt)
            if treat_as_local:
                return naive.replace(tzinfo=zone).astimezone(timezone.utc)
            return naive.replace(tzinfo=timezone.utc)
        except ValueError:
            continue

    try:
        # True ISO with offset
        return datetime.fromisoformat(text.replace("Z", "+00:00")).astimezone(timezone.utc)
    except ValueError:
        return None


def _as_float(value: Any) -> float | None:
    if value is None:
        return None
    try:
        return float(value)
    except (TypeError, ValueError):
        return None


def _as_int(value: Any) -> int | None:
    if value is None:
        return None
    try:
        return int(value)
    except (TypeError, ValueError):
        return None


def parse_bolus_entries(graph_data: dict[str, Any], zone: ZoneInfo | None = None) -> list[BolusEntry]:
    series = graph_data.get("series") or graph_data
    if not isinstance(series, dict):
        return []
    items = series.get("deliveredBolus") or []
    entries: list[BolusEntry] = []
    for item in items:
        if not isinstance(item, dict):
            continue
        time_raw = item.get("timestamp", item.get("x"))
        ts = parse_timestamp(time_raw, zone)
        units = _as_float(item.get("insulinDelivered", item.get("y", item.get("value"))))
        if ts is None or units is None or units <= 0:
            continue
        is_manual = item.get("isManual")
        entries.append(
            BolusEntry(
                timestamp=ts,
                units=units,
                carbs_input=_as_float(item.get("carbsInput")),
                insulin_on_board=_as_float(item.get("insulinOnBoard")),
                blood_glucose_input=_as_int(item.get("bloodGlucoseInput")),
                blood_glucose_source=item.get("bloodGlucoseInputSource"),
                correction_units=_as_float(item.get("insulinRecommendationForCorrection")),
                carb_units=_as_float(item.get("insulinRecommendationForCarbs")),
                is_manual=bool(is_manual) if isinstance(is_manual, bool) else False,
                device_name=item.get("deviceName"),
            )
        )
    return entries


def _argmax_mode(auto: float, manual: float, limited: float) -> PumpMode:
    if auto >= manual and auto >= limited:
        return PumpMode.AUTO
    if manual >= limited:
        return PumpMode.MANUAL
    return PumpMode.LIMITED


def parse_pump_mode(statistics: dict[str, Any]) -> PumpStatistics | None:
    if not isinstance(statistics, dict):
        return None

    auto = statistics.get("op5PumpModeAutomaticPercentage")
    manual = statistics.get("op5PumpModeManualPercentage")
    limited = statistics.get("op5PumpModeLimitedPercentage")
    if auto is not None or manual is not None or limited is not None:
        auto_f = _as_float(auto) or 0.0
        manual_f = _as_float(manual) or 0.0
        limited_f = _as_float(limited) or 0.0
        return PumpStatistics(
            mode=_argmax_mode(auto_f, manual_f, limited_f),
            auto_percentage=auto_f,
            manual_percentage=manual_f,
            limited_percentage=limited_f,
            hypo_protect_percentage=_as_float(statistics.get("op5PumpModeHypoProtectPercentage")) or 0.0,
            total_insulin_per_day=_as_float(statistics.get("totalInsulinPerDay")),
            basal_percentage=_as_float(statistics.get("basalPercentage")),
            bolus_percentage=_as_float(statistics.get("bolusPercentage")),
            average_bg=_as_float(statistics.get("averageBg")),
            in_range_percentage=_as_float(statistics.get("inRangePercentage")),
            carbs_per_day=_as_float(statistics.get("carbsPerDay")),
        )

    modes = statistics.get("pumpModes") or statistics.get("pump_modes")
    if isinstance(modes, dict):
        auto_f = _as_float(modes.get("auto", modes.get("automated"))) or 0.0
        manual_f = _as_float(modes.get("manual")) or 0.0
        limited_f = _as_float(modes.get("limited")) or 0.0
        if auto_f or manual_f or limited_f:
            return PumpStatistics(
                mode=_argmax_mode(auto_f, manual_f, limited_f),
                auto_percentage=auto_f,
                manual_percentage=manual_f,
                limited_percentage=limited_f,
            )
    return None


def parse_devices(device_data: dict[str, Any], zone: ZoneInfo | None = None) -> list[DeviceInfo]:
    if not isinstance(device_data, dict):
        return []
    devices_raw = device_data.get("devices") or []
    devices: list[DeviceInfo] = []
    for item in devices_raw:
        if not isinstance(item, dict):
            continue
        name = item.get("displayName") or item.get("shortDisplayName") or "Unknown"
        props_raw = item.get("properties") or {}
        props = {str(k): str(v) for k, v in props_raw.items()} if isinstance(props_raw, dict) else {}
        devices.append(
            DeviceInfo(
                name=name,
                brand=item.get("brand"),
                model=item.get("model"),
                serial_number=item.get("serialNumber"),
                device_type=item.get("type"),
                last_sync=parse_timestamp(item.get("lastSyncTimestamp"), zone),
                properties=props,
            )
        )
    return devices
